packages/build/src/plugins/mutations.js

```diff
@@ -49,55 +49,25 @@ const serializeKeys = function (keys) {
   return keys.map(String).join('.')
 }
 
-// When setting `build.command`, `build.commandOrigin` is set to "plugin"
-const setBuildCommandOrigin = function (priorityConfig) {
-  // eslint-disable-next-line fp/no-mutation, no-param-reassign
-  priorityConfig.build.commandOrigin = 'plugin'
-}
-
-const setTopFunctionsDirectory = function (priorityConfig, value, keys) {
-  setFunctionsDirectory(priorityConfig, value)
-  setFunctionsCatchAll(priorityConfig, value, keys)
-}
-
-// When settings `functions.{propName}`, we also set `functions.*.{propName}`,
-// emulating the normalization performed by `@netlify/config`.
-const setFunctionsCatchAll = function (priorityConfig, value, keys) {
-  const lastKey = keys[keys.length - 1]
+// `functionsDirectory` is created by `@netlify/config`.
+// We denormalize it to `functions.directory` which is user-facing.
+const denormalizeFunctionsDirectory = function (priorityConfig, functionsDirectory) {
   // eslint-disable-next-line fp/no-mutation, no-param-reassign
-  priorityConfig.functions['*'][lastKey] = value
-}
-
-// Several configuration properties can be used to specify the functions directory.
-// `netlifyConfig.functionsDirectory` is the normalized property which must be set.
-// We allow plugin authors to set any of the other properties for convenience.
-const setFunctionsDirectory = function (priorityConfig, value) {
+  priorityConfig.functions = priorityConfig.functions || {}
   // eslint-disable-next-line fp/no-mutation, no-param-reassign
-  priorityConfig.functionsDirectory = value
+  priorityConfig.functions.directory = functionsDirectory
 }
 
 // List of properties that are not read-only.
 const MUTABLE_PROPS = {
-  'build.command': { lastEvent: 'onPreBuild', handler: setBuildCommandOrigin },
-  'build.commandOrigin': { lastEvent: 'onPreBuild' },
-  'build.functions': { lastEvent: 'onBuild', handler: setFunctionsDirectory },
+  'build.command': { lastEvent: 'onPreBuild' },
+  'build.functions': { lastEvent: 'onBuild' },
   'build.publish': { lastEvent: 'onPostBuild' },
   'build.edge_handlers': { lastEvent: 'onPostBuild' },
-  functionsDirectory: { lastEvent: 'onBuild' },
+  functionsDirectory: { lastEvent: 'onBuild', denormalize: denormalizeFunctionsDirectory },
```

Comment:
The only `denormalize()` we need is for `functionsDirectory`, which is denormalized to `functions.directory`.
In fact, we could just force plugins to update `functions.directory`. However, I kept it as an example of denormalization so we can test the denormalization logic since it will be used for future configuration properties.